File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/AlgorithmInfo.kt
package com.jetbrains.signatureverifier.serialization

import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.ASN1ObjectIdentifier
import org.bouncycastle.asn1.ASN1Primitive
import org.bouncycastle.asn1.DLSequence
import org.bouncycastle.asn1.x509.AlgorithmIdentifier
import org.bouncycastle.operator.DefaultAlgorithmNameFinder

// additionalValue is to be investigated, for now it is just StringInfo or null
@Serializable
data class AlgorithmInfo(
  val name: String,
  val additionalValue: TextualInfo? = null,
  val algorithmIdentifier: TextualInfo
) : EncodableInfo {
  constructor(signatureAlgorithm: AlgorithmIdentifier) : this(
    DefaultAlgorithmNameFinder().getAlgorithmName(signatureAlgorithm.algorithm as ASN1ObjectIdentifier),
    signatureAlgorithm.parameters?.let {
      TextualInfo.getInstance(
        it
      )
    },
    TextualInfo.getInstance(signatureAlgorithm.algorithm)
  )

  private fun toDLSequence(): DLSequence {
    val list = mutableListOf(algorithmIdentifier.toPrimitive())
    if (additionalValue != null) {
      list.add(additionalValue.toPrimitive())
    }
    return list.toDLSequence()
  }

  override fun toPrimitive(): ASN1Primitive = toDLSequence().toASN1Primitive()
}

File: net/JetBrains.Serialization/src/Dataholders/AlgorithmInfo.cs
using System;
using System.Collections.Generic;
using JetBrains.Annotations;
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Org.BouncyCastle.Asn1.X509;
using Org.BouncyCastle.Cms;

namespace JetBrains.Serialization;

[JsonObject(MemberSerialization.OptIn)]
public class AlgorithmInfo : IEncodableInfo
{
  [JsonProperty("name")] private String _name;

  [JsonProperty("additionalValue")] private IEncodableInfo? _additionalValue = null;

  [JsonProperty("algorithmIdentifier")] private TextualInfo _algorithmIdentifier;

  [JsonConstructor]
  public AlgorithmInfo(string name, TextualInfo algorithmIdentifier)
  {
    _name = name;
    _algorithmIdentifier = algorithmIdentifier;
  }

  public AlgorithmInfo(AlgorithmIdentifier signatureAlgorithm)
  {
    _name = SerializationUtils.AlgorithmNameFromId(signatureAlgorithm.Algorithm);
    _additionalValue = signatureAlgorithm.Parameters?.ToAsn1Object()?.ToEncodableInfo();
    _algorithmIdentifier = TextualInfo.GetInstance(signatureAlgorithm.Algorithm);
  }

  public Asn1Encodable ToPrimitive() => new List<IEncodableInfo?> { _algorithmIdentifier, _additionalValue }
    .ToPrimitiveDerSequence()
    .ToAsn1Object();
}